domain/application/state/application.go

```diff
@@ -874,30 +880,168 @@ INSERT INTO k8s_service (*) VALUES ($cloudService.*)
 		if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
 			return errors.Errorf(
 				"querying cloud service for application %q and provider id %q: %w", applicationName, providerID, err)
+		} else if errors.Is(err, sqlair.ErrNoRows) {
+			// Nothing already exists so create a new net node and placeholder
+			// link layer device for the cloud service.
+			nodeUUID, err := st.insertCloudServiceNetNode(ctx, tx)
+			if err != nil {
+				return errors.Errorf("creating cloud service net node for application %q: %w", applicationName, err)
+			}
+			serviceInfo.NetNodeUUID = nodeUUID.String()
+
+			uuid, err := uuid.NewUUID()
+			if err != nil {
+				return errors.Capture(err)
+			}
+			serviceInfo.UUID = uuid.String()
+			if err := tx.Query(ctx, insertStmt, serviceInfo).Run(); err != nil {
+				return errors.Capture(err)
+			}
 		}
-		if err == nil {
-			return nil
+
+		if len(sAddrs) > 0 {
+			// Retrieve the link layer device UUID for the service.
+			var lldUUIDStr string
+			err = tx.Query(ctx, queryLinkLayerDeviceFromServiceStmt, serviceInfo).Get(&linkLayerDeviceUUID)
+			if err != nil && !errors.Is(err, sqlair.ErrNoRows) {
+				return errors.Errorf("querying cloud service link layer device for application %q: %w", applicationName, err)
+			} else if errors.Is(err, sqlair.ErrNoRows) {
+				// Ensure the address link layer device is inserted.
+				lldUUID, err := st.insertCloudServiceDevice(ctx, tx, applicationName, serviceInfo.NetNodeUUID)
+				if err != nil {
+					return errors.Errorf("inserting cloud service link layer device for application %q: %w", applicationName, err)
+				}
+				lldUUIDStr = lldUUID.String()
+			} else {
+				lldUUIDStr = linkLayerDeviceUUID.UUID
+			}
+
+			// Before inserting the new addresses, we need to remove any existing
+			// ones for the given application and provider id.
+			if err := st.deleteCloudServiceAddresses(ctx, tx, appUUID, providerID); err != nil {
+				return errors.Capture(err)
+			}
+			if err := st.insertCloudServiceAddresses(ctx, tx, applicationName, lldUUIDStr, sAddrs); err != nil {
+				return errors.Capture(err)
+			}
 		}
```

Comment:
This should be a separate method. `UpsertCloudServices` is already big.
For cloud containers the `upsertCloudContainerAddress` does the logic - we could introduce a `upsertCloudServiceAddresses` to contain the logic here